### DISCUSSION TITLE
O que deveríamos saber hoje?

### DISCUSSION THREAD

**Discussion by icaromh** [frontendbr/forum#114] [Discussão]
Pessoal, 
fui convidado pra dar uma "palestra" em formato de conversa mesmo, lá na faculdade. E o tema é: 

**"O que deveríamos saber hoje?"**

A ideia é dar uma força pra quem tá começando e orientar no estilo: Cara, para de criar layouts em tabela, estuda sobre flex-box. Ou, mano larga esse php 5.3 e tenta estudar um pouco sobre o que tem de novo. React, Laravel, saber(ao menos) o que é docker e por aí vai.

Como front, gosto de ter uma noção geral de todo funcionamento do sistema. Até pra poder criar minhas próprias arquiteturas pra freelas e tal.

Minha dúvida é: O que vocês acham essencial saber hoje? 

Edit.: Nada melhor que a comunidade dar a direção :+1:

**Comment by yurimatheus** [comment]
Acho essencial saber:
1. Mobile (Ionic / AngularJS / React / etc);
2. Real Time (Socket.io)
3. NoSQL (Redis e MongoDB)
4. Técnicas de Gameficação
5. Internet das Coisas

**Comment by diessica** [comment]
Esse é um tópico que só pode ser levantado com especificidade, senão todos vão interpretar de uma forma diferente. Quem é o "nós" da sua pergunta? Onde essa sabedoria vai ser aplicada? 

Dependendo das respostas, eu discordo dos itens citados pelo @yurimatheus. Eu rebateria: Eles são essenciais para quem? Para que fim?

Ionic, Angular e React **NÃO** são coisas que _deveríamos_ (ênfase!) saber hoje. Existem várias habilidades que devemos ter atualmente, e elas não competem somente à técnica – e, quando competem, a avaliação técnica a ser feita não é essa.

**Comment by lagden** [comment]
Tem que saber programar, conhecer bem sua linguagem e o resto é praticar!!
Framework em um semana você aprende!!

**Comment by heliomsolivas** [comment]
No caso do front-end acredito que continua sendo o html semântico, css3, javascript, git, dessa base tu segue um caminho, é complicado o termo "deveriamos", acho que tudo depende da área.

**Comment by icaromh** [comment]
@diessica entendi, vou tentar melhorar nisso. :smile:

A ideia da conversa é levantar algo que se assemelhe ao [Radar](https://www.thoughtworks.com/pt/radar/) para a pessoa consiga se manter um pouco atualizada. 

@lagden concordo, em partes, não se aprende em uma semana. Dá pra entender e tudo, mas fluência em um framework demora até tu cair nas armadilhas deles :p

A ideia do direcionamento não é exatamente na carreira. Até porquê são só 30 minutos de conversa. Mas sim tecnicamente, expor um pouco das tecnologias que estão sendo discutidas hoje e, explicar por cima o que é cada um delas e porquê estão surgindo.

**Comment by diessica** [comment]
É que o essencial é essencial, e saber o essencial não quer dizer que você está atualizado. Ainda não entendi a proposta do tópico. Você fez **três** perguntas:
1. O que deveríamos saber hoje?
2. O que a pessoa precisa para se manter atualizada hoje?
3.  O que vocês acham essencial saber hoje?

As respostas são diferentes para as três. Na primeira, podemos abordar até metodologias ágeis. Na segunda, posso responder "saber se virar psicológica e tecnicamente muito bem sem um mentor", e, na terceira, acho que podemos chegar em uma resposta produtiva.

#### O que vocês acham essencial saber hoje?

O essencial técnico é HTML semântico e acessível, CSS metodológico e modular, JavaScript modular e seguindo Design Patterns, noções de UX... Isso é essencial para **qualquer** projeto.

Agora, se você quer falar de _estar na crista da onda_, podemos falar de React, Immutable.js, Webpack, npm, JavaScript testado... Mas daí isso vai variar demais, já não vale para qualquer projeto.

**Comment by afonsopacifer** [comment]
@icaromh, ano passado eu cheguei a falar em algumas faculdades sobre o mesmo assunto,
no caso minha talk era "Front-end no mundo real". 

Talvez os slides te deem uma direção:
https://speakerdeck.com/afonsopacifer/fron-end-no-mundo-real

Geralmente em faculdades os paranaues de Front são desconhecidos.. acho que uma visão geral da realidade seja o que esse publico precise ;)

AH. e avisa o pessoal que existe uma comunidade <3

**Comment by lagden** [comment]
@icaromh você discorda porque não domina sua linguagem... acho até uma semana muito!!
A base e o domínio da sua linguagem é tudo o que você precisa! Like a Bruce Lee!!

**Comment by baltazarparra** [comment]
De tudo citado acima, sem dúvidas conhecimento sólido em HTML, CSS e JavaScript é o mais importante. O mercado já está saturado de "Desenvolvedor jQuery" e HTML sujo com classes de bootstrap.

**Comment by felipeuntill** [comment]
Na minha opinião como adendo ao que a @diessica comentou, os requisitos não são globais, cada nicho de mercado possui suas especifidades, algumas empresas são tradicionais e não há quem modifique as tecnologias.

**Comment by danguilherme** [comment]
Concordo plenamente com a @diessica. É **essencial** pra mim saber comunicação em tempo real? Técnicas de gameficação?! Essencial é eu conseguir escrever um bom código, de fácil manutenção e leitura. Quando eu precisar implementar comunicação em tempo real aprendo o socket.io em uma semana, como o @lagden disse.

**Comment by lagden** [comment]
@danguilherme isso aee!!! pressão!! uma semana!!! Socket.io é suave...
depois você adiciona uns clusters para distribuir a carga!! gameficação é tesão...

**Comment by kazzkiq** [comment]
Eu trabalhei em uma startup onde tive a oportunidade de conviver com vários profissionais fenomenais (back-end ou full-stack). Aprendi muita coisa e progredi muito como profissional com a ajuda deles, mas o melhor aprendizado que tive com eles pode ser resumido em uma palavra: **desenvoltura**.

Um desenvolvedor que possui capacidade de lidar com novas tecnologias de acordo com a demanda do projeto, que não tem medo de aprender coisa nova, e que tem uma base sólida de conhecimento na sua área (lógica de programação, algoritmos, design patterns, etc) se sai bem em praticamente qualquer ambiente.

Todos os caras que entraram na empresa tiveram que fazer manutenção em um framework que eles nunca tinham ouvido falar (era da própria empresa). Além disso, tinha toda uma fábrica de software com 60 programadores cobrando novas versões do tal framework, e reclamando de bugs que ainda não tinham sido resolvidos. Pressão insana, mas os caras tinham muito conhecimento nas suas respectivas áreas e mesmo pegando uma tecnologia nova, em poucas semanas já estavam se virando bem, e em poucos meses já tinham virado referência na empresa para aquela ferramenta.

O que eu (e um monte de gente nas respostas aí em cima) quero dizer com este exemplo é: Não se prenda à uma tecnologia. Vale muito mais a pena investir em uma base sólida de conhecimento que possa ser aplicada em várias tecnologias, do que focar unicamente na "moda do momento" e perder valor de mercado depois de 6 meses, quando outra moda entrar no lugar.

Um profissional que tenha conhecimento sólido de JavaScript e que não tenha medo de aprender novas tecnologias caso necessário, é muito mais valioso do que outro que só sabe programar se for com a IDE dele, no padrão dele, com o framework dele.

É muito melhor ser perito em JavaScript do que em jQuery. Ou React. Ou Angular. Ou Ember. Se você focar o seu conhecimento na base, você pode aplicar ela para qualquer uma dessas outras tecnologias. Isso te torna alguém com desenvoltura. E pessoas com desenvoltura geralmente dão um jeito de resolver os problemas, não criá-los. Não precisa nem dizer o valor que uma empresa vê em um profissional assim né. 😉

**Comment by felipefialho** [comment]
De total acordo com a @diessica. 

O que sempre precisamos estar alinhados é com a BASE sagrada do desenvolvimento web, ou seja: **HTML**, **CSS** e **JavaScript** e noções de **UX**.

O resto das ferramentas e conceitos dependem muito mais da empresa que está contratando (ou você trabalhando) do que como conceitos globais.

E uma vez escrevi uma frase no meu site que cabe muito bem nesse cenário:

> Importante lembrar que ferramentas, aplicativos e todas essas maravilhas modernas, são apenas facilitadores para que possamos entregar o melhor trabalho possível. Se apegue mais ao conceito, e menos nas ferramentas, porque normalmente, elas são passageiras.

**Comment by felquis** [comment]
Front-end Job Interview Questions é um ótimo lugar para se aprender o Essencial, visto que, as perguntas fazem parte da base do desenvolvimento web, sem poluição de libs e frameworks.

-> https://github.com/h5bp/Front-end-Developer-Interview-Questions

Também é muito bom entender o que acontece quando você digita google.com na url e tecla enter 
-> https://github.com/alex/what-happens-when

Mais um link interessante dessa vez de uma empresa privada -> https://www.toptal.com/javascript/interview-questions

Enfim... acho que a galera respondeu muito bem também anteriormente.. Parabéns galera!

**Comment by felipefialho** [comment]
Nossa, que legal esse segundo repositório @felquis, não conhecia. Valeu por compartilhar.

**Comment by brunopulis** [comment]
Boa discurssão. 

Acredito que todo frontend deve conhecer com propriedade a **trindade sagrada: HTML, CSS, JS** como foi dito pelo @LFeh e é bastante interessante saber sobre Usabilidade e especialmente sobre acessibilidade web, visto que, as aplicações hoje estão cada vez mais interativas. 

E coisas essenciais que na minha opinião acho que deve saber:
- GIT;
- Metodologia de CSS (BEM, OOCSS, RSCSS);
- VanillaJs;

**Comment by dougaraujos** [comment]
Pra não chover no molhado, além de tudo que foi dito, acho essencial o conhecimento e, se possível, domínio sobre como seu trabalho (code) é interpretado nos browsers.

Metodologias, frameworks e afins, tendem mais para preferências e/ou exigências. Contudo, é imprescindível garantir que a execução que não compete mais à você (interpretação dos browsers, cache, protocolos, compressão, etc) seja a melhor possível.

Indico um post do Paul Irish sobre funcionamento do browser, (traduzido em 2011): http://www.html5rocks.com/pt/tutorials/internals/howbrowserswork/

**Comment by leandroruel** [comment]
tem que saber também fazer um código indentado e manutenível, para não f**\* com os colegas que mexerão em seu código também.

**Comment by diessica** [comment]
Quando eu postei as perguntas:

> Quem é o "nós" da sua pergunta? Onde essa sabedoria vai ser aplicada?

Eu estava falando sobre os diferentes perfis de desenvolvedores front-end, e os diferentes projetos que nos envolvemos. Explorei um pouco mais esse tópico [neste artigo](https://diessi.ca/blog/a-fantastica-diversidade-do-front-end/), a fim de ampliarmos nossa visão sobre o nosso próprio trabalho! :)

**Comment by icaromh** [comment]
Pessoal, acabo por me convencer haha :smile: 

Obrigado pela força no assunto, vou repensar o que apresentar lá e levar pra mim também.

**Comment by bernardodiasc** [comment]
Muito bom o artigo @diessica!

Acho que praticamente tudo ja foi dito. :)

Se o fim é o browser, então boa base em HTML, CSS e JS são requisitos obrigatórios. Metodologias e ferramentas são importantes mas são secundárias visto que não são necessariamente necessárias e variam muito de caso em caso.

Além disso, acho importante hoje em dia e arrisco a incluir na categoria de fundamental: GIT, interpretação de texto e inglês. :)

E dados estruturados!!! Dados estão por toda parte e influenciam tudo por essas bandas virtuais.

**Comment by felipefialho** [comment]
Boa @icaromh. Consegue contar pra gente como foi depois? 

E o @bernardodiasc tem total razão, eu incluiria **GIT, interpretação de texto e inglês**, nós nunca falamos disso como requisitos mínimos, mas são.

Uma pessoa que não sabe interpretar um texto, pode colocar em risco todo o projeto, já que tem grandes chances de desenvolver completamente diferente do que tinha sido planejado.

**Comment by bernardodiasc** [comment]
Sobre interpretação de texto, pensa se voce fosse empregador e colocasse uma questão desse tipo pro entrevistado resolver, e se o resultado da interpretação e por sua vez a elaboração desta fosse nonsense, ou mesmo com erros ortográficos. Eu não contrataria nunca, não precisaria nem fazer os testes técnicos. Lembro bem quando participava da comunidade de webdev no google+, cada figura que aparecia que não dava nem pra entender o que estava escrito. Certa ocasião sugeri ao caboclo melhorar a escrita pra poder ser levado a sério, mas acho que ele ficou ofendido com a sugestão, na verdade eu não havia conseguido entender nada da dúvida da pessoa, o que ele havia escrito não fazia nenhum sentido kkkkkkk

Sobre GIT, pude observar que praticamente todas as vagas não colocam como requisito, embora seja com grande chances a ferramenta adotada pela empresa. Sou do tempo do FTP, sei que muitas agencias xing-ling ainda usam, mas FTP pode ser legal pra estagiários talvez, pra quem ta realmente começando (não não, to zuando). Mesmo assim, num curso que apliquei com novatos no ramo, GIT foi a primeira técnica que ensinei, antes de HTML, em seguida ensinei Markdown, que tem a ver com entender o que significa transformar uma estrutura de dados em outra e é fácil o suficiente pra novatos entenderem, a abstração desse conceito é amplamente utilizada quando a complexidade do código (falando basicamente de html, css e js aqui) aumenta.

**Comment by fernandoadms** [comment]
Na área de Front-end acredito que a pessoa deve saber, no mínimo isto:
- HTML5/ CSS3
- Javascript
- Semântico
- Alguma ferramenta de controle de versionamento (GIT/SVN...)

O resto eu acredito que vai funcionar para alguns profissionais e de acordo com cada projeto. As vezes o Angular atende a sua necessidade, as vezes não.

Por isso acredito que o mais importante é saber o feijão com arroz:
- HTML/CSS
- Javascript

E o resto é ler e se manter antenado sempre! (y)

**Comment by lagden** [comment]
@felquis agora eu sei o que acontece when the "x" key is pressed... 🦄

**Comment by rubenmarcus** [comment]
E sobre as vagas que fazem testes do tipo , escreva uma chamada ajax X , fale das funções do wordpress, e escreva qual cada uma faz, qual dessas funções jquery faz isso ou isso, vocês concordam que estudar na nossa profissão,é saber detalhadamente o que cada coisa faz, ou ter um bom raciocinio,pra saber implementar as melhores soluções sem trazer soluções decoradas?

Falo isso,porque todos empregos que eu consegui, foi por meu portfolio e trabalhos e nunca por testes.

E sou bem da linha de ter uma base solida e ferramentas voce aprende em semanas e meses, eu pelo menos nunca tive problema com isso.. 

Coisas que a uns anos atras eu considerava back-end e pra mim era complicado como manutenção de servidor, mexer mais a fundo em banco de dados etc,etc eu to aprendendo pq estou sendo obrigado a aprender..

Eu vejo pelo menos que a minha linha de curva sempre muda devido a isso, a curiosidade e a necessidade, nunca devido ao mercado.. vocês acham que se pautar pelo mercado é uma boa linha de pensamento?

**Comment by leandroruel** [comment]
só pra complementar aqui que, além de todas as habilidades já citadas aqui, existe uma básica que todo desenvolvedor deve saber, e que ninguém mencionou aqui, que é... saber fazer um miojo. afinal, dev que é dev ou pede fast food ou faz miojo. rs

**Comment by felipefialho** [comment]
@leandroruel Cara, eu pelo contrário, amo cozinhar, treinar receitas e etc. É um hobbie do carario para relaxar.

O @Mendrone, o @matmarsiglio e eu estávamos até falando de fazer um MeetupCook, haha.

**Comment by mtmr0x** [comment]
@Mendrone @LFeh E eu já avisei o Mendrone que eu estou terminando meu apto lá pra irmos lá fazer. Me da umas semanas pra chegar minha cozinha nova e vamos! 

Se der certo, a gente expande e faz um MeetupCook hahhahaha

**Comment by elfiservice** [comment]
Vale a pena da o Check aqui by Udacity    ✅ 
[Habilidades de um desenvolvedor front-end](https://br.udacity.com/checklist-desenvolvedor-front-end?utm_source=vero&utm_medium=email&utm_content=a&utm_campaign=2018-05-25_BR%20Ebook%20Checklist%20de%20front-end&utm_term=Newsletter&vero_id=9572678855&vero_conv=Z3_qSuIADu6XuicUr7QGYTFmricer7zFuSEeuWcIK9ypm1oKh8QgvkP7RZ2b8WFbEjsn_I6R6jnX0dro6zLHbwhrhnDjy4gj)
